speaker/layer2_controller.go

```diff
@@ -25,12 +28,69 @@ import (
 
 type layer2Controller struct {
 	announcer *layer2.Announce
+	myNode    string
 }
 
 func (c *layer2Controller) SetConfig(log.Logger, *config.Config) error {
 	return nil
 }
 
+// usableNodes returns all nodes that have at least one fully ready
+// endpoint on them.
+func usableNodes(eps *v1.Endpoints) []string {
+	usable := map[string]bool{}
+	for _, subset := range eps.Subsets {
+		for _, ep := range subset.Addresses {
+			if ep.NodeName == nil {
+				continue
+			}
+			if _, ok := usable[*ep.NodeName]; !ok {
+				// Only set true if nothing else has expressed an
+				// opinion. This means that false will take precedence
+				// if there's any unready ports for a given endpoint.
+				usable[*ep.NodeName] = true
+			}
+		}
+
+		for _, ep := range subset.NotReadyAddresses {
+			if ep.NodeName == nil {
+				continue
+			}
+			usable[*ep.NodeName] = false
+		}
+	}
+
+	ret := make([]string, 0, len(usable))
```

Comment:
Total nit, but since you may use less capacity based on the `ok` state in the loop below, would it be best to just let append do its thing and grow this as needed, eg `var ret []string`?

Reply:
Good call, done.